tools/selection_tools.cpp

```diff
@@ -255,6 +255,65 @@ void CloseSelection(Mesh* obj)
 	CloseSelection (obj->selector());
 }
 
+void RestrictSelectionToSubset(Mesh* obj, int si)
```

Comment:
There is a lot of code duplication in that method. There could be an additional template method which performs the restriction for one element type. You'd then call that method here for 'Vertex', 'Edge', ...
Also - why do you store a list of selected elements, deselect all and reselect those elements. Instead you should simply deselect the ones which are not in the specified subset.